### DISCUSSION TITLE
Improve markdown code blocks: adding language and user attributes

### DISCUSSION THREAD

**Discussion by EmNudge** [withastro/roadmap#491] [Proposal]
### Body

# Summary

Add back the parsed language from code blocks into the resulting markup.

# Background & Motivation

In earlier versions of Astro, code blocks included the language and did not have inline styling. Currently, code blocks do not have the language included in the resulting markup.

![image](https://user-images.githubusercontent.com/24513691/219561076-3f965853-e2d3-45bf-8795-b4cd3d1b4bda.png)

While the code is styled as if it is javascript, the `js` string was not added to any part of the resulting markup. I had code that relied on being able to parse what kind of language the code block was. If it was JavaScript, special considerations could be made. 

# Goals

* Add back language to resulting markup. Perhaps as a class name in the `<pre>` tag.
* Allow for additional metadata applied to code blocks. Extend remark syntax.

# Example

Astro code:
```
```js
console.log('hi');
\```
```

Resulting markup:
```html
<pre class="astro-code js">
<code>
<!-- some highlighted markup -->
</code>
</pre>
```

**Comment by delucis** [comment]
Hey @EmNudge — the inline styling is provided by [Shiki](https://github.com/shikijs/shiki) which is what Astro uses to process code blocks by default.

However, you can still choose to switch to Prism for syntax highlighting and retain the older behaviour you preferred. See the [Syntax highlighting guide in the Astro Docs](https://docs.astro.build/en/guides/markdown-content/#syntax-highlighting) for more details.

**Reply by EmNudge** [reply]
Yeah, the background color specifically seems to be from Shiki, not Astro's shiki remark plugin. Thanks for the comment!

The removal of language is easily fixed within Astro however. I'm going to edit this discussion to be more specific to the language attribute.

**Comment by bluwy** [comment]
This seems to be the same as https://github.com/withastro/roadmap/discussions/276

**Comment by naiyerasif** [comment]
If you're OK with using a Rehype plugin, I wrote [one](https://github.com/Microflash/rehype-starry-night) that highlights the code blocks with [starry-night](https://github.com/wooorm/starry-night) and adds the language information (amongst other things). 

It supports applying additional metadata on code blocks (e.g., highlighting lines, prompt indicators, captions, etc). There's no API to add your own metadata though.
